### ISSUE TITLE
sample of coaps_server was tested failed on qemu

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1605]
**_Reported by jiequan wu:_**

Configuration
==========
<Arch> x86
Bug detailed description
===================
sample of coaps_server was tested failed on qemu
==================
Test steps:
```
net-tools:
	Terminal 1:
	$ ./loop-socat.sh

	Terminal 2:
	$ sudo ./loop-slip-tap.sh
Zephyr:
	$ cd coaps_server
	$ make pristine && make run
Libcoap:
        $ ./autogen.sh
        $ ./configure --disable-shared
        $ make all
	$ cd examples
	$ ./coap-client -m get coaps://[2001:db8::1]/test -u Client_identity -k passwd
```

Qemu output:
```
To exit from QEMU enter: 'CTRL+a, x'
[QEMU] CPU: qemu32
Packet without payload
No handler for such request (-22)
Packet without payload
No handler for such request (-22)
Packet without payload
No handler for such request (-22)
Packet without payload
No handler for such request (-22)
Packet without payload
No handler for such request (-22)
 mbedtls_ssl_read returned -0x7780
```
Host output:
```
linux@linux-desktop:~/latest/libcoap/examples$ ./coap-client -m get coaps://[2001:db8::1]/test -u Client_identity -k passwd
v:1 t:CON c:GET i:4c5b {} [ ]
decrypt_verify(): found 24 bytes cleartext
```

(Imported from Jira ZEP-1754)

**Comment by nashif** [comment]
**_by Jukka Rissanen:_**

In the description you probably meant to say

Zephyr:
	$ cd coaps_server
	$ make pristine && make run

as there is no such directory as coap_server any more.

Anyway, assigning to Sergio.

**Comment by nashif** [comment]
**_by Sergio Rodriguez:_**

From the  of it seems that DTLS handshake, encryption and decryption works correctly, maybe has the cause is some CoAP parsing issues, assigning to Vinicius

**Comment by nashif** [comment]
**_by Vinicius Costa Gomes:_**

The "problem" was that a few method handlers were considering a request without payload to be errors. It was only noticed now because, the behavior of zoap_packet_get_payload() is more strict. See if the proposed change makes sense.

**Comment by nashif** [comment]
**_by jiequan wu:_**

The rc2(commit 46a38ee4601b4a) also has this issue.

**Comment by nashif** [comment]
**_by Anselmo L. S. Melo:_**

_jiequan wu_ , please check if https://gerrit.zephyrproject.org/r/#/c/11877/ fixes this.
It was not part of rc2, so the problem was expected with that tag - The same applies to rc3, as this patch got merged after that tag as well.

**Comment by nashif** [comment]
**_by jiequan wu:_**

I tested  https://gerrit.zephyrproject.org/r/#/c/11877/  already fixes this issue. But this patch is not merged on master branch.
```
libcoap/examples$ ./coap-client -m get coaps://[2001:db8::1]/test -u Client_identity -k passwd
v:1 t:CON c:GET i:afb5 {} [ ]
decrypt_verify(): found 24 bytes cleartext
decrypt_verify(): found 32 bytes cleartext
Type: 0
Code: 1
MID: 44981
```
Qemu output:
```
To exit from QEMU enter: 'CTRL+a, x'
[QEMU] CPU: qemu32
qemu-system-i386: warning: Unknown firmware file in legacy mode: genroms/multiboot.bin

*******
type: 0 code 1 id 44981
*******
 mbedtls_ssl_read returned -0x7780
```

**Comment by nashif** [comment]
**_by Anselmo L. S. Melo:_**

It is on v1.7-branch, thus might be in the upcoming 1.7-rc4: https://gerrit.zephyrproject.org/r/gitweb?p=zephyr.git;a=commit;h=6e135d0006fec916898ac00788db1927c757cba3

**Comment by nashif** [comment]
**_by jiequan wu:_**

If possible merged on master, I will close this issue.

**Comment by nashif** [comment]
**_by Anselmo L. S. Melo:_**

The patch is already where it should be (v1.7-branch) to be part of the 1.7 release, which is the target of QA verification at this moment.

I understand your concern about this getting integrated into master (which is already receiving patches for 1.8). AFAIU this patch will get merged as example of other patches currently in net branch.

_Vinicius Costa Gomes_ , _Kuo-Lang Tseng_ , _Anas Nashif_

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

_jiequan wu_ , Yes, _Anselmo L. S. Melo_  is right. This will be merged to master as part of the regular merge from net-> master.

**Comment by nashif** [comment]
**_by Anselmo L. S. Melo:_**

Fixed for 1.7.
This patch will be in the master branch with the next net branch merge.
